ARCs/arc-0004.md

```diff
@@ -0,0 +1,498 @@
+SHA-512/256 hash (in hex): 8aa3b61f0f1965c3a1cbfa91d46b24e54c67270184ff89dc114e877b1753254a
+Method selector (in hex): 8aa3b61f
+```
+
+
+#### Method Description
+
+A method description provides further information about a method
+beyond its signature. This description is encoded in JSON and consists
+of a method's name, description, arguments (their types, names, and
+descriptions), and return type (and possible description for the
+return type). From this structure, the method's signature and selector
+can be calculated. The Algorand SDKs will provide convenience
+functions to calculate signatures and selectors from such JSON files.
+
+These details will enable high-level languages and dapps/wallets to
+properly encode arguments and call methods, as well as offer users a
+rich description for a method and its arguments. This description can
+populate UIs in dapps, wallets, and explorers with description of
+parameters, as well as populate information about methods in IDEs for
+HLLs.
+
+The JSON structure for such an object is:
+
+```
+interface Method {
+  name: string,
+  desc?: string,
+  args: Array<{ name: string, type: string, desc?: string }>,
+  returns?: { type: string, desc?: string }
+}
+```
+
+For example:
+
+```
+{
+  "name": "add",
+  "desc": "Calculate the sum of two 64-bit integers",
+  "args": [
+    { "name": "a", "type": "uint64", "desc": "..." },
+    { "name": "b", "type": "uint64", "desc": "..." }
+  ],
+  "returns": { "type": "uint128", "desc": "..." }
+}
+```
+
+
+### Interfaces
+
+An interface is a logically grouped set of methods. All methods in an
+interface must be unique (specifically, they must have unique method
+selectors). Method names in Interfaces **MUST NOT** begin with an
+underscore. A smart contract implements an interface if it supports
+all of the methods from that interface. A smart contract may implement
+zero, one, or multiple interfaces.
+
+#### Interface Description
+
+An interface description is a JSON object containing the JSON
+descriptions for each of the methods in the interface.
+
+The JSON structure for such an object is:
+
+```
+interface Interface {
+  name: string,
+  methods: Array<Method>
+}
+```
+
+For example:
+
+```
+{
+  "name": "Calculator",
+  "methods": [
+    { "name": "add", "args": [...], ... },
+    { "name": "multiply", "args": [...], ... }
+  ]
+}
+```
+
+### Contracts
+
+An contract is the complete set of methods that an app implements. It
+is similar to an interface, but may include further details about the
+concrete implementation. All methods in an contract must be unique
+(specifically, they must have unique method selectors). Method names
+in Contracts **MAY** begin with underscore, but these names are
+reserved for use by this, or suceeding specifications.
+
+A contract may contain two special methods, `_optIn` and
+`_closeOut`. These methods describe how contracts should opt-in, or
+close-out such apps.  They may have any arguments or return values,
```

Comment:
Right, we specify nothing about ClearState right now, in part because that is already a "last ditch" option for accounts to rid themselves of state.  There's not much left to say.